src/coreclr/jit/optimizer.cpp

```diff
@@ -4575,6 +4575,276 @@ PhaseStatus Compiler::optUnrollLoops()
+    // Detect via the block weight as that will be high when inside a loop.
+    if (block->getBBWeight(this) > BB_UNITY_WEIGHT)
+    {
+        return false;
+    }
+
+    // Does the block end by branching via a JTRUE after a compare?
+    if (block->bbJumpKind != BBJ_COND || block->NumSucc() != 2)
+    {
+        return false;
+    }
+
+    // Verify the test block ends with a conditional that we can manipulate.
+    GenTree* last = block->lastStmt()->GetRootNode();
+    noway_assert(last->OperIs(GT_JTRUE));
+    if (!last->AsOp()->gtOp1->OperIsCmpCompare() || (last->gtFlags & GTF_SIDE_EFFECT) != 0 ||
+        (last->AsOp()->gtOp1->gtFlags & GTF_COLON_COND) != 0)
+    {
+        return false;
+    }
+
+    BasicBlock* middleBlock = block;
+    GenTree*    asgNode     = nullptr;
+    Statement*  asgStmt     = nullptr;
+    bool        foundSelect = false;
+
+    // Check the block is followed by a block or chain of blocks that only contain NOPs and
+    // a single ASG statement. The destination of the final block must point to the same as the
+    // true path of the JTRUE block.
+    bool foundMiddle = false;
+    while (!foundMiddle)
+    {
+        middleBlock = middleBlock->bbNext;
+        noway_assert(middleBlock != nullptr);
+
+        if (middleBlock->bbNext == block->bbJumpDest)
+        {
+            // This is our final middle block.
+            foundMiddle = true;
+        }
+
+        // Make sure there is only one block which jumps to the middle block,
+        // and the middle block is not the start of a TRY block or an exception handler.
+        noway_assert(!fgCheapPredsValid);
+        if (middleBlock->NumSucc() != 1 || middleBlock->bbJumpKind != BBJ_NONE ||
+            middleBlock->bbPreds->flNext != nullptr || middleBlock->bbCatchTyp != BBCT_NONE ||
+            bbIsTryBeg(middleBlock) || bbIsHandlerBeg(middleBlock) ||
+            (middleBlock->bbFlags & BBF_DONT_REMOVE) != 0)
+        {
+            return false;
+        }
+
+        // Can all the nodes within the middle block be made to conditionally execute?
+        for (Statement* const stmt : middleBlock->Statements())
+        {
+            GenTree* tree = stmt->GetRootNode();
+            switch (tree->gtOper)
+            {
+                case GT_ASG:
+                {
+                    GenTree* op1 = tree->gtGetOp1();
+                    GenTree* op2 = tree->gtGetOp2();
+
+                    // Only one per assignment per block can be conditionally executed.
+                    // Ensure the destination of the assign is a local variable with integer type,
+                    // and the nodes of the assign won't cause any additional side effects.
+                    if (asgNode != nullptr || !op1->OperIs(GT_LCL_VAR) || !varTypeIsIntegralOrI(op1) ||
+                        (op1->gtFlags & GTF_SIDE_EFFECT) != 0 || (op2->gtFlags & GTF_SIDE_EFFECT) != 0)
+                    {
+                        return false;
+                    }
```

Comment:
Looking at this again, we definitely need to check `GTF_ORDER_SIDEFF` here, and not just for volatile loads. The JIT may remove side-effect flags based on upstream checks and doing if-conversion in those cases is not safe. For example:
```csharp
[MethodImpl(MethodImplOptions.NoInlining)]
private static int Foo(C c)
{
    int i = 123;
    if (c != null)
        i = c.F;
    return i;
}

private class C
{
    public int F;
}
```
The JIT will see:
```scala
***** BB01
STMT00000 ( 0x000[E-] ... 0x002 )
N003 (  5,  4) [000002] -A------R--                         ▌  ASG       int    $VN.Void
N002 (  3,  2) [000001] D------N---                         ├──▌  LCL_VAR   int    V01 loc0         d:2 $VN.Void
N001 (  1,  1) [000000] -----------                         └──▌  CNS_INT   int    123 $42

***** BB01
STMT00001 ( 0x003[E-] ... 0x004 )
N004 (  5,  5) [000006] -----------                         ▌  JTRUE     void   $VN.Void
N003 (  3,  3) [000005] J------N---                         └──▌  EQ        int    $100
N001 (  1,  1) [000003] -----------                            ├──▌  LCL_VAR   ref    V00 arg0         u:1 $80
N002 (  1,  1) [000004] -----------                            └──▌  CNS_INT   ref    null $VN.Null

------------ BB02 [006..00D), preds={BB01} succs={BB03}

***** BB02
STMT00003 ( 0x006[E-] ... 0x00C )
N006 (  8,  7) [000012] -A--GO--R--                         ▌  ASG       int    $281
N005 (  3,  2) [000011] D------N---                         ├──▌  LCL_VAR   int    V01 loc0         d:4 $VN.Void
N004 (  4,  4) [000010] n---GO-----                         └──▌  IND       int    <l:$102, c:$103>
N003 (  2,  2) [000014] -------N---                            └──▌  ADD       byref  $180
N001 (  1,  1) [000009] -----------                               ├──▌  LCL_VAR   ref    V00 arg0         u:1 (last use) $80
N002 (  1,  1) [000013] -----------                               └──▌  CNS_INT   long   8 field offset Fseq[F] $140
```

Notice that there are no side-effect flags on `[000010]` (G = `GTF_GLOB_REF`, O = `GTF_ORDER_SIDEFF`) but doing if-conversion here is incorrect.

I suspect this is the cause of most of the failures.

Reply:
>I suspect this is the cause of most of the failures.

I think that was the case - there were some IND nodes.